src/include/wiredtiger.in

```diff
@@ -4647,6 +4664,179 @@ struct __wt_file_handle {
 };
 #endif /* !defined(SWIG) */
 
+#ifdef DOXYGEN
+/*! WT_SHARED_STORAGE::open_object flags: create a new object */
+#define	WT_SS_OPEN_CREATE	0x0
+/*! WT_SHARED_STORAGE::open_object flags: open is read-only */
+#define	WT_SS_OPEN_READONLY	0x0
+#else
+/* AUTOMATIC FLAG VALUE GENERATION START */
+#define	WT_SS_OPEN_CREATE	0x1u
+#define	WT_SS_OPEN_READONLY	0x2u
+/* AUTOMATIC FLAG VALUE GENERATION STOP */
+#endif
+
+/*!
+ * The interface implemented by applications to provide a shared system
+ * implementation. This documentation refers to "object" and "bucket"
+ * to mean a "file-like object" and a "container of objects", respectively.
+ *
+ * <b>Thread safety:</b> WiredTiger may invoke methods on the WT_SHARED_STORAGE
+ * interface from multiple threads concurrently. It is the responsibility of
+ * the implementation to protect any shared data.
+ *
+ * Applications register implementations with WiredTiger by calling
+ * WT_CONNECTION::add_shared_storage.  See @ref custom_shared_storages for more
+ * information.
+ *
+ * @snippet ex_shared_storage.c WT_SHARED_STORAGE register
+ */
+struct __wt_shared_storage {
+	/*!
+	 * Return a location handle from a location string.
+         * A location string may encode a bucket name, or the equivalent for this
+         * shared storage system, authorization information for that bucket,
+         * naming prefixes to be used for objects in that bucket, etc.
+	 *
+	 * @errors
+	 *
+	 * @param shared_storage the WT_SHARED_STORAGE
+	 * @param session the current WiredTiger session
+	 * @param location the location string
+	 * @param[out] location_handle the allocated handle
+	 */
+	int (*ss_location_handle)(WT_SHARED_STORAGE *shared_storage,
+	    WT_SESSION *session, const char *location, void **location_handle);
+
+	/*!
+	 * Free a location handle created by WT_SHARED_STORAGE::ss_location_handle.
+	 *
+	 * @errors
+	 *
+	 * @param shared_storage the WT_SHARED_STORAGE
+	 * @param session the current WiredTiger session
+	 * @param location_handle the handle to be freed.
+	 */
+	int (*ss_location_handle_free)(WT_SHARED_STORAGE *shared_storage,
+	    WT_SESSION *session, void *location_handle);
+
+	/*!
+	 * Return a list of object names for the given location.
+	 *
+	 * @errors
+	 *
+	 * @param shared_storage the WT_SHARED_STORAGE
+	 * @param session the current WiredTiger session
+	 * @param location_handle the location to list
+	 * @param[out] objlist the method returns an allocated array of
```

Comment:
Will this "list everything" method be a problem if we have a bucket with a huge number of objects?  Obviously this gets back to some of the questions we've been discussing about how buckets will be used.

Reply:
Yes, that's true.  Perhaps we need some sort of streamed result.  You know the `WT_FILE_SYSTEM` api has a method to list a single object.  Perhaps that is also useful.

Reply:
The "list single" method was added in WT-3750 as a performance path for logging. If we ever want to extract out internally, we probably want or need something similar.

Reply:
I've modified the list API to 1) add a prefix (this is available in the file system interface) and 2) set a limit for how many objects to return.  The file system interface has a separate API call to list just one, that is achieved here by setting the limit to 1.